### PR TITLE
[WIP] Simple session prototype

### PR DESCRIPTION
**Pull request by allouis**
This is a hacked together simple session prototype (no tests, questionable use of library).

These changes _completely_ remove the existing auth stuff in reality we'd want to wrap both mechanisms up I believe?

This changes will only work with https://github.com/TryGhost/Ghost-Admin/pull/1043

If you run this locally you'll wanna add a top level property of `"session-secret": "something"` to your config.json

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,6 +43,7 @@
     "chalk": "1.1.3",
     "cheerio": "0.22.0",
     "compression": "1.7.2",
+    "connect-session-knex": "1.4.0",
```
Comment (allouis) [inline package.json]: This was quick and easy session persistence - I checked the code and it seems decent. but we might not wanna actually use this.

File: core/server/services/auth/session.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    return User.check({
+        email: username,
+        password
+    }).then(function (user) {
+        req.session.user_id = user.id;
+        req.session.save(function (err) {
+            if (err) {
+                return res.send(500);
+            }
+            res.send(201, {});
```
Comment (allouis) [inline core/server/services/auth/session.js]: This was a funny lil thing where ember was expecting JSON and going nuts if it didn't get any.

File: core/server/services/auth/session.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    if (!req.body) {
+        return next(new BadRequestError());
+    }
+    const {username, password} = req.body;
+    return User.check({
+        email: username,
+        password
+    }).then(function (user) {
+        req.session.user_id = user.id;
+        req.session.save(function (err) {
```
Comment (allouis) [inline core/server/services/auth/session.js]: Not 100% sure we need to call save here

File: core/server/services/auth/session.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    next(new UnauthorizedError('Missing credentials'));
+};
+
+const getSession = session({
+    store: new KnexSessionStore(db),
+    secret: config.get('session-secret'),
+    resave: false,
+    saveUninitialised: false,
+    cookie: {
+        maxAge: 184 * 7 * 24 * 60 * 60 * 1000, // number of days in second half of year
```
Comment (allouis) [inline core/server/services/auth/session.js]: :beer: :question:

File: core/server/web/api/routes.js [hunk 5]
```diff
@@ -30,6 +30,9 @@ module.exports = function apiRoutes() {
     // ## CORS pre-flight check
     apiRouter.options('*', cors);
 
+    apiRouter.use(auth.session.getSession);
```
Comment (allouis) [inline core/server/web/api/routes.js]: This ensures `req.session` exists

File: core/server/web/api/routes.js [hunk 6]
```diff
@@ -30,6 +30,9 @@ module.exports = function apiRoutes() {
...
+    apiRouter.use(auth.session.getUser);
```
Comment (allouis) [inline core/server/web/api/routes.js]: This sets `req.user` to either `{user_id: <id>}` or `null`

File: core/server/web/api/middleware.js [hunk 7]
```diff
@@ -29,9 +25,7 @@ module.exports.authenticatePublic = [
  * Authentication for private endpoints
  */
 module.exports.authenticatePrivate = [
-    auth.authenticate.authenticateClient,
-    auth.authenticate.authenticateUser,
-    auth.authorize.requiresAuthorizedUser,
+    auth.session.ensureUser,
```
Comment (allouis) [inline core/server/web/api/middleware.js]: Errors if `req.user.id` doesn't exist or doesn't map to a user

File: core/server/services/auth/session.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return res.send(204);
+    });
+};
+
+const getUser = function getUser(req, res, next) {
+    if (!req.session || !req.session.user_id) {
+        req.user = null;
+        return next();
+    }
+    User.findOne({id: req.session.user_id})
```
Comment (allouis) [inline core/server/services/auth/session.js]: This probably wants to check use isn't disabled?

**Comment by kirrg001** [comment]
Can we close this PR? 🤔

**Comment by allouis** [comment]
Yes :)

**Pull request closed without merging**